pipelines/rbsa/tasks/undiscount_gas.py

```diff
@@ -0,0 +1,78 @@
+import logging
+import pandas as pd
+from generics import task as t
+import json
+
+logger = logging.getLogger('LCTK_APPLICATION_LOGGER')
+
+class UndiscountGas(t.Task):
+    """ 
+    This class is used to group sites into 3 digit zip codes
+    """
+    def __init__(self, name):
+        super().__init__(self)
+        self.name = name
+
+        self.input_artifact_enduse_loads = 'enduse_loads.csv'
+        self.input_artifact_gas_fraction = 'GAS_FRACTIONS.json'
+        self.input_artifact_zip_zone_map = 'ZIP_ZONE_MAP.json'
+        self.my_data_files = [self.input_artifact_enduse_loads] 
+
+        self.output_artifact_total_loads = 'total_loads.csv'
+        
+        self.task_function = self._task
+
+    def _get_data(self):
+        return self.load_data(self.my_data_files)
+        
+    def _task(self):
+        data_map = self._get_data()
+
+        self.df = data_map[self.input_artifact_enduse_loads]
+
+        # TODO replace with json reader
+        with open('config/GAS_FRACTIONS.json') as json_file:  
+            self.gas_fraction = json.load(json_file) 
+        with open('config/ZIP_ZONE_MAP.json') as json_file:  
+            self.zip_zone_map = json.load(json_file) 
+
+        # output dataframe initialization
+        initialization = True
+
+        zipcodes = self.df.zipcode.unique()
+
+        for zipcode in zipcodes:
+
+            zipcode_df = self.df.loc[self.df.zipcode == zipcode]
+            zipcode_df = zipcode_df.reset_index()
+
+            zone = self.zip_zone_map['mapping'][str(zipcode)]
+            electric_percentage = self.gas_fraction['electrification'][zone]   
+
+            # add gas fraction
+            for enduse in electric_percentage.keys():
+                zipcode_df[enduse] = zipcode_df[enduse]/electric_percentage[enduse]
+        
+            # output dataframe 
+            if initialization:
+                total_loads = zipcode_df
```

Comment:
I hadn't thought about this, but why not just initialize the `total_loads` variable before the for loop as an empty df and then just always run the append below. Seems like you'd be able to get rid of the `initialization` variable

Reply:
Also, I think something similar applies to the other tasks. Worth updating if that's the case.

Reply:
I adjust the df format during the first loop and wanted to use that version as the df's setup. Do you think that's an issue? I can change all if so.

Reply:
I'm not sure what the data is looking like, but I can run it and see if there's some room for simplification. But it currently works and it's fine.

One thought is, instead of this:
```
     for zipcode in zipcodes:

            zipcode_df = self.df.loc[self.df.zipcode == zipcode]

            zone = self.zip_zone_map['mapping'][str(zipcode)]
            electric_percentage = self.gas_fraction['electrification'][zone]   

            # add gas fraction
            for enduse in electric_percentage.keys():
                zipcode_df[enduse] = zipcode_df[enduse]/electric_percentage[enduse]

            # output dataframe 
            if initialization:
                total_loads = zipcode_df
                initialization = False
            else:
                total_loads = total_loads.append(zipcode_df)

        self.validate(total_loads)
        self.save_data({self.output_artifact_total_loads: total_loads})
```

you could do this:
```
    total_loads = None
    for zipcode in zipcodes:
            zipcode_df = self.df.loc[self.df.zipcode == zipcode]
            zone = self.zip_zone_map['mapping'][str(zipcode)]
            electric_percentage = self.gas_fraction['electrification'][zone]   

            # add gas fraction
            for enduse in electric_percentage.keys():
                zipcode_df[enduse] = zipcode_df[enduse]/electric_percentage[enduse]

            # output dataframe 
            if total_loads:
                total_loads = total_loads.append(zipcode_df)
            else:
                total_loads = zipcode_df
            
        self.validate(total_loads)
        self.save_data({self.output_artifact_total_loads: total_loads})
```

Reply:
The indentations aren't correct on the snippet above, but you get the gist.

If it doesn't make sense right now it's fine - we can circle back later since this is an optimization that is not worth spending too much time on.

Reply:
That makes sense, let me give it a quick test